transport/transport_test.go

```diff
@@ -382,6 +382,45 @@ func setUpWithNoPingServer(t *testing.T, copts ConnectOptions, done chan net.Con
 	return tr
 }
 
+// TestInflightStreamClosing tests that closing in-flight stream trigger a server
+// to send GoAway to the client with in-flight stream.
+func TestInflightStreamClosing(t *testing.T) {
+	serverConfig := &ServerConfig{
+		KeepaliveParams: keepalive.ServerParameters{
+			MaxConnectionIdle: 2 * time.Second,
+		},
```

Comment:
If we rely on keepalive related goaways, we don't test the code you wrote `s.write(recvMsg{err: err}) `.
The test should be a simple reproduction of your prod issue; a goroutine is blocked on stream.Read() while clientconn.Close() is called, with your change in place this goroutine should unblock.